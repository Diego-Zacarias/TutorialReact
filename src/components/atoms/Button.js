import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

export const ButtonColors = {
  default: "default",
  primary: "primary",
  danger: "danger",
};

export const ButtonVariants = {
  default: "default",
  outlined: "outlined",
  link: "link",
};

const colorMain = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.main;
    case ButtonColors.danger:
      return theme.colors.danger.main;
    default:
      return "#909090";
  }
};
const darkColor = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.dark;
    case ButtonColors.danger:
      return theme.colors.danger.dark;
    default:
      return "#5a6268";
  }
};

const colorText = (props) => props.theme.colors.primary.text;

const Button = styled.button`
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: ${colorMain};
  border: 2px solid ${colorMain};
  color: ${colorText};

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:hover:enabled {
    background-color: ${darkColor};
    border-color: ${darkColor};
  }
`;

const ButtonOutlined = styled(Button)`
  background-color: transparent;
  color: ${colorMain};

  &:hover:enabled {
    background-color: transparent;
    color: ${darkColor};
  }
`;

const ButtonLink = styled(Button)`
  background-color: transparent;
  border-color: transparent;
  color: ${colorMain};
  padding-left: 0;
  padding-right: 0;

  &:hover:enabled {
    background-color: transparent;
    border-color: transparent;
    color: ${darkColor};
  }
`;

const ButtonWrapper = (props) => {
  switch (props.variants) {
    case ButtonVariants.outlined:
      return <ButtonOutlined {...props} />;
    case ButtonVariants.link:
      return <ButtonLink {...props} />;
    default:
      return <Button {...props} />;
  }
};

ButtonWrapper.defaultProps = {
  type: "buton",
  children: undefined,
  color: "default",
  variants: "default",
};

ButtonWrapper.propTypes = {
  type: PropTypes.string,
  children: PropTypes.node,
  color: PropTypes.oneOf(Object.values(ButtonColors)),
  variants: PropTypes.oneOf(Object.values(ButtonVariants)),
};

export default ButtonWrapper;
