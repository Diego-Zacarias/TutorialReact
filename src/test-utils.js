import React from "react";
import { render } from "@testing-library/react";
import ThemeProvider from "styles/ThemeProvider";
import { MemoryRouter } from "react-router-dom";

const TestProvider = ({ children }) => (
  <MemoryRouter>
    <ThemeProvider theme={"ocean"}>{children}</ThemeProvider>
  </MemoryRouter>
);

export * from "@testing-library/react";

const customRender = (ui, options) =>
  render(ui, { wrapper: TestProvider, ...options });

export { customRender as render };
