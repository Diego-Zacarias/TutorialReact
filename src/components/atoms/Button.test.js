import React from "react";
import { render, fireEvent } from "test-utils";

import Button, { ButtonColors, ButtonVariants } from "./Button";

test("renders a text", () => {
  const { getByText } = render(<Button>Click here</Button>);

  expect(getByText("Click here")).toBeInTheDocument();
});

test("trigger event on click", () => {
  const handleClick = jest.fn();
  const { getByRole } = render(<Button onClick={handleClick} />);

  fireEvent.click(getByRole("button"));
  expect(handleClick).toBeCalled();
});

test.each(Object.values(ButtonColors).map((item) => [item]))(
  "renders with color %s",
  (color) => {
    const { asFragment } = render(<Button color={color} />);

    expect(asFragment()).toMatchSnapshot();
  }
);

test.each(Object.values(ButtonVariants).map((item) => [item]))(
  "renders with variant %s",
  (variant) => {
    const { asFragment } = render(<Button variant={variant} />);

    expect(asFragment()).toMatchSnapshot();
  }
);
