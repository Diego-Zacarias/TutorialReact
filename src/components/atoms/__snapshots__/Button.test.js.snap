// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with color danger 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ff3d00;
  border: 2px solid #ff3d00;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #b22a00;
  border-color: #b22a00;
}

<button
    class="c0"
    color="danger"
    type="buton"
  />
</DocumentFragment>
`;

exports[`renders with color default 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #909090;
  border: 2px solid #909090;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="buton"
  />
</DocumentFragment>
`;

exports[`renders with color primary 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #12f3da;
  border: 2px solid #12f3da;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #01d1b5;
  border-color: #01d1b5;
}

<button
    class="c0"
    color="primary"
    type="buton"
  />
</DocumentFragment>
`;

exports[`renders with variant default 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #909090;
  border: 2px solid #909090;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="buton"
  />
</DocumentFragment>
`;

exports[`renders with variant link 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #909090;
  border: 2px solid #909090;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="buton"
  />
</DocumentFragment>
`;

exports[`renders with variant outlined 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #909090;
  border: 2px solid #909090;
  color: #212121;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="buton"
  />
</DocumentFragment>
`;
